// fireorm/lib/src/helpers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBatch = exports.runTransaction = exports.GetBaseRepository = exports.getBaseRepository = exports.GetCustomRepository = exports.getCustomRepository = exports.GetRepository = exports.getRepository = void 0;
var MetadataUtils_1 = require("./MetadataUtils");
var BaseFirestoreRepository_1 = require("./BaseFirestoreRepository");
var FirestoreTransaction_1 = require("./Transaction/FirestoreTransaction");
var FirestoreBatch_1 = require("./Batch/FirestoreBatch");
// TODO: return transaction repo. Is it needed?!?
function _getRepository(entityConstructorOrPath, repositoryType) {
    var metadataStorage = MetadataUtils_1.getMetadataStorage();
    if (!metadataStorage.firestoreRef) {
        throw new Error('Firestore must be initialized first');
    }
    var collection = metadataStorage.getCollection(entityConstructorOrPath);
    var isPath = typeof entityConstructorOrPath === 'string';
    var collectionName = typeof entityConstructorOrPath === 'string'
        ? entityConstructorOrPath
        : entityConstructorOrPath.name;
    // TODO: create tests
    if (!collection) {
        var error = isPath
            ? "'" + collectionName + "' is not a valid path for a collection"
            : "'" + collectionName + "' is not a valid collection";
        throw new Error(error);
    }
    var repository = metadataStorage.getRepository(collection.entityConstructor);
    if (repositoryType === 'custom' && !repository) {
        throw new Error("'" + collectionName + "' does not have a custom repository.");
    }
    // If the collection has a parent, check that we have registered the parent
    if (collection.parentEntityConstructor) {
        var parentCollection = metadataStorage.getCollection(collection.parentEntityConstructor);
        if (!parentCollection) {
            throw new Error("'" + collectionName + "' does not have a valid parent collection.");
        }
    }
    if (repositoryType === 'custom' || (repositoryType === 'default' && repository)) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return new (repository === null || repository === void 0 ? void 0 : repository.target)(entityConstructorOrPath);
    }
    else {
        return new BaseFirestoreRepository_1.BaseFirestoreRepository(entityConstructorOrPath);
    }
}
function getRepository(entityConstructorOrPath) {
    return _getRepository(entityConstructorOrPath, 'default');
}
exports.getRepository = getRepository;
/**
 * @deprecated Use getRepository. This will be removed in a future version.
 */
exports.GetRepository = getRepository;
function getCustomRepository(entityOrPath) {
    return _getRepository(entityOrPath, 'custom');
}
exports.getCustomRepository = getCustomRepository;
/**
 * @deprecated Use getCustomRepository. This will be removed in a future version.
 */
exports.GetCustomRepository = getCustomRepository;
function getBaseRepository(entityOrPath) {
    return _getRepository(entityOrPath, 'base');
}
exports.getBaseRepository = getBaseRepository;
/**
 * @deprecated Use getBaseRepository. This will be removed in a future version.
 */
exports.GetBaseRepository = getBaseRepository;
var runTransaction = function (executor) { return __awaiter(void 0, void 0, void 0, function () {
    var metadataStorage;
    return __generator(this, function (_a) {
        metadataStorage = MetadataUtils_1.getMetadataStorage();
        if (!metadataStorage.firestoreRef) {
            throw new Error('Firestore must be initialized first');
        }
        return [2 /*return*/, metadataStorage.firestoreRef.runTransaction(function (t) { return __awaiter(void 0, void 0, void 0, function () {
                var tranRefStorage, result;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            tranRefStorage = new Set();
                            return [4 /*yield*/, executor(new FirestoreTransaction_1.FirestoreTransaction(t, tranRefStorage))];
                        case 1:
                            result = _a.sent();
                            tranRefStorage.forEach(function (_a) {
                                var entity = _a.entity, path = _a.path, propertyKey = _a.propertyKey;
                                var record = entity;
                                record[propertyKey] = getRepository(path);
                            });
                            return [2 /*return*/, result];
                    }
                });
            }); })];
    });
}); };
exports.runTransaction = runTransaction;
var createBatch = function () {
    var metadataStorage = MetadataUtils_1.getMetadataStorage();
    if (!metadataStorage.firestoreRef) {
        throw new Error('Firestore must be initialized first');
    }
    return new FirestoreBatch_1.FirestoreBatch(metadataStorage.firestoreRef);
};
exports.createBatch = createBatch;
